maven-compiler-plugin arguments missing when compiling projects with Javac

```
!ENTRY org.eclipse.jdt.core 4 0 2024-09-04 14:25:15.616
!MESSAGE compilation failed
!STACK 0
com.sun.tools.javac.processing.AnnotationProcessingError: java.lang.IllegalArgumentException: 'tlibVersion' is missing
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:1033)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:937)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1265)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1378)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1271)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:947)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:897)
	at org.eclipse.jdt.internal.javac.JavacCompiler.compile(JavacCompiler.java:162)
	at org.eclipse.jdt.internal.core.builder.AbstractImageBuilder.compile(AbstractImageBuilder.java:414)
	at org.eclipse.jdt.internal.core.builder.BatchImageBuilder.compile(BatchImageBuilder.java:211)
	at org.eclipse.jdt.internal.core.builder.AbstractImageBuilder.compile(AbstractImageBuilder.java:345)
	at org.eclipse.jdt.internal.core.builder.BatchImageBuilder.build(BatchImageBuilder.java:79)
	at org.eclipse.jdt.internal.core.builder.JavaBuilder.buildAll(JavaBuilder.java:286)
	at org.eclipse.jdt.internal.core.builder.JavaBuilder.build(JavaBuilder.java:192)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:1077)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:296)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:352)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:441)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:444)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:555)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:503)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:585)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:207)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:300)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.IllegalArgumentException: 'tlibVersion' is missing
	at org.apache.struts.annotations.taglib.apt.TagAnnotationProcessor.checkOptions(TagAnnotationProcessor.java:223)
	at org.apache.struts.annotations.taglib.apt.TagAnnotationProcessor.process(TagAnnotationProcessor.java:74)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:1021)
	... 26 more
```

The sample project is https://github.com/apache/struts.git. The javac bits failed to compile it due to some annotation processing args missing.

In the original pom.xml, it specifies the arguments as follows:
```
<plugin>
                        <groupId>org.apache.maven.plugins</groupId>
                        <artifactId>maven-compiler-plugin</artifactId>
                        <configuration>
                            <compilerArgs>
                                <arg>-Auri=/struts-tags</arg>
                                <arg>-AtlibVersion=${tlib.version}</arg>
                                <arg>-AjspVersion=2.0</arg>
                                <arg>-AshortName=s</arg>
                                <arg>-AdisplayName=Struts Tags</arg>
                                <arg>-AoutFile=${basedir}/target/classes/META-INF/struts-tags.tld</arg>
                                <arg>-Adescription="To make it easier to access dynamic data;
                                    the Apache Struts framework includes a library of custom tags.
                                    The tags interact with the framework's validation and
                                    internationalization features;
                                    to ensure that input is correct and output is localized.
                                    The Struts Tags can be used with JSP FreeMarker or Velocity."</arg>
                                <arg>-AoutTemplatesDir=${basedir}/src/site/resources/tags</arg>
                            </compilerArgs>
                        </configuration>
                        <executions>
                            <execution>
                                <phase>compile</phase>
                                <goals>
                                    <goal>compile</goal>
                                </goals>
                            </execution>
                        </executions>
                    </plugin>
```

And when m2e imports the projects, it converts these arguments as annotation processing into the setting org.eclipse.jdt.apt.core.prefs.
```
eclipse.preferences.version=1
org.eclipse.jdt.apt.aptEnabled=true
org.eclipse.jdt.apt.genSrcDir=target/generated-sources/annotations
org.eclipse.jdt.apt.genTestSrcDir=target/generated-test-sources/test-annotations
org.eclipse.jdt.apt.processorOptions/description="To make it easier to access dynamic data;\n                                    the Apache Struts framework includes a library of custom tags.\n                                    The tags interact with the framework's validation and\n                                    internationalization features;\n                                    to ensure that input is correct and output is localized.\n                                    The Struts Tags can be used with JSP FreeMarker or Velocity."
org.eclipse.jdt.apt.processorOptions/displayName=Struts Tags
org.eclipse.jdt.apt.processorOptions/jspVersion=2.0
org.eclipse.jdt.apt.processorOptions/outFile=...\\struts\\core/target/classes/META-INF/struts-tags.tld
org.eclipse.jdt.apt.processorOptions/outTemplatesDir=...\\struts\\core/src/site/resources/tags
org.eclipse.jdt.apt.processorOptions/shortName=s
org.eclipse.jdt.apt.processorOptions/tlibVersion=6.0
org.eclipse.jdt.apt.processorOptions/uri=/struts-tags
```

Lastly, when javac compiler compile the projects, it does not pick such ap arguments.

--- a/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java
+++ b/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacUtils.java
@@ -19,6 +19,7 @@ import java.util.LinkedHashSet;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Queue;
 import java.util.Set;
 import java.util.function.Predicate;
@@ -166,6 +167,11 @@ public class JavacUtils {
 			} catch (JavaModelException ex) {
 				ILog.get().error(ex.getMessage(), ex);
 			}
+		}
+
+		Map<String, String> processorOptions = ProcessorConfig.getProcessorOptions(javaProject);
+		for (Entry<String, String> processorOption : processorOptions.entrySet()) {
+			options.put("-A" + processorOption.getKey() + "=" + processorOption.getValue(), Boolean.toString(true));
 		}
 	}
 

--- a/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/ProcessorConfig.java
+++ b/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/ProcessorConfig.java
@@ -1,0 +1,150 @@
+/*******************************************************************************
+ * Copyright (c) 2005, 2018 BEA Systems, Inc. and others
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ * 
+ * Copied from eclipse.jdt.core/org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/util/AptConfig.java
+ *
+ * Contributors:
+ *    jgarms@bea.com, wharley@bea.com - initial API and implementation
+ *    het@google.com - Bug 423254 - There is no way to tell if a project's factory path is different from the workspace default
+ *******************************************************************************/
+
+package org.eclipse.jdt.internal.javac;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IWorkspaceRoot;
+import org.eclipse.core.resources.ProjectScope;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.ILog;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
+import org.eclipse.core.runtime.preferences.IScopeContext;
+import org.eclipse.core.runtime.preferences.InstanceScope;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.osgi.service.prefs.BackingStoreException;
+
+public class ProcessorConfig {
+	private static final String APT_PLUGIN_ID = "org.eclipse.jdt.apt.core"; //$NON-NLS-1$
+	private static final String APT_STRING_BASE = "org.eclipse.jdt.apt"; //$NON-NLS-1$ 1$
+	private static final String APT_PROCESSOROPTIONS = APT_STRING_BASE + ".processorOptions"; //$NON-NLS-1$
+	private static final String APT_NULLVALUE = APT_STRING_BASE + ".NULLVALUE"; //$NON-NLS-1$
+
+	/** regex to identify substituted token in path variables */
+	private static final String PATHVAR_TOKEN = "^%[^%/\\\\ ]+%.*"; //$NON-NLS-1$
+	/** path variable meaning "workspace root" */
+	private static final String PATHVAR_ROOT = "%ROOT%"; //$NON-NLS-1$
+	/** path variable meaning "project root" */
+	private static final String PATHVAR_PROJECTROOT = "%PROJECT.DIR%"; //$NON-NLS-1$
+
+	public static Map<String, String> getProcessorOptions(IJavaProject jproj) {
+		Map<String, String> rawOptions = getRawProcessorOptions(jproj);
+		// map is large enough to also include the programmatically generated options
+		Map<String, String> options = new HashMap<>(rawOptions.size());
+
+		// Resolve path metavariables like %ROOT%
+		for (Map.Entry<String, String> entry : rawOptions.entrySet()) {
+			String resolvedValue = resolveVarPath(jproj, entry.getValue());
+			String value = (resolvedValue == null) ? entry.getValue() : resolvedValue;
+			options.put(entry.getKey(), value);
+		}
+
+		return options;
+	}
+
+	private static Map<String, String> getRawProcessorOptions(IJavaProject jproj) {
+		Map<String, String> options = new HashMap<>();
+
+		// Fall back from project to workspace scope on an all-or-nothing basis,
+		// not value by value. (Never fall back to default scope; there are no
+		// default processor options.) We can't use IPreferencesService for this
+		// as we would normally do, because we don't know the names of the keys.
+		IScopeContext[] contexts;
+		if (jproj != null) {
+			contexts = new IScopeContext[] { new ProjectScope(jproj.getProject()), InstanceScope.INSTANCE };
+		} else {
+			contexts = new IScopeContext[] { InstanceScope.INSTANCE };
+		}
+		for (IScopeContext context : contexts) {
+			IEclipsePreferences prefs = context.getNode(APT_PLUGIN_ID);
+			try {
+				if (prefs.childrenNames().length > 0) {
+					IEclipsePreferences procOptionsNode = context.getNode(APT_PLUGIN_ID + "/" + APT_PROCESSOROPTIONS); //$NON-NLS-1$
+					if (procOptionsNode != null) {
+						for (String key : procOptionsNode.keys()) {
+							String nonNullVal = procOptionsNode.get(key, null);
+							String val = APT_NULLVALUE.equals(nonNullVal) ? null : nonNullVal;
+							options.put(key, val);
+						}
+						break;
+					}
+				}
+			} catch (BackingStoreException e) {
+				ILog.get().error("Unable to load annotation processor options", e); //$NON-NLS-1$
+			}
+		}
+		return options;
+	}
+
+	/**
+	 * If the value starts with a path variable such as %ROOT%, replace it with the
+	 * absolute path.
+	 * 
+	 * @param value the value of a -Akey=value command option
+	 */
+	private static String resolveVarPath(IJavaProject jproj, String value) {
+		if (value == null) {
+			return null;
+		}
+		// is there a token to substitute?
+		if (!Pattern.matches(PATHVAR_TOKEN, value)) {
+			return value;
+		}
+		IPath path = new Path(value);
+		String firstToken = path.segment(0);
+		// If it matches %ROOT%/project, it is a project-relative path.
+		if (PATHVAR_ROOT.equals(firstToken)) {
+			IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
+			IResource proj = root.findMember(path.segment(1));
+			if (proj == null) {
+				return value;
+			}
+			// all is well; do the substitution
+			IPath relativePath = path.removeFirstSegments(2);
+			IPath absoluteProjPath = proj.getLocation();
+			IPath absoluteResPath = absoluteProjPath.append(relativePath);
+			return absoluteResPath.toOSString();
+		}
+
+		// If it matches %PROJECT.DIR%/project, the path is relative to the current
+		// project.
+		if (jproj != null && PATHVAR_PROJECTROOT.equals(firstToken)) {
+			// all is well; do the substitution
+			IPath relativePath = path.removeFirstSegments(1);
+			IPath absoluteProjPath = jproj.getProject().getLocation();
+			IPath absoluteResPath = absoluteProjPath.append(relativePath);
+			return absoluteResPath.toOSString();
+		}
+
+		// otherwise it's a classpath-var-based path.
+		String cpvName = firstToken.substring(1, firstToken.length() - 1);
+		IPath cpvPath = JavaCore.getClasspathVariable(cpvName);
+		if (cpvPath != null) {
+			IPath resolved = cpvPath.append(path.removeFirstSegments(1));
+			return resolved.toOSString();
+		} else {
+			return value;
+		}
+	}
+}

